modules/consensus/accept.go

```diff
@@ -55,6 +56,73 @@ func (cs *ConsensusSet) validateHeader(tx dbTx, b types.Block) error {
 	return cs.blockValidator.ValidateBlock(b, minTimestamp, parent.ChildTarget, parent.Height+1)
 }
 
+// checkTarget returns true if the header's ID meets the given target.
+func checkHeaderTarget(h types.BlockHeader, target types.Target) bool {
+	blockHash := h.ID()
+	return bytes.Compare(target[:], blockHash[:]) >= 0
+}
+
+// validateHeader does some early, low computation verification on the header
+// to determine if the block should be downloaded. Callers should not assume
+// that validation will happen in a particular order.
+func (cs *ConsensusSet) validateHeader(tx dbTx, h types.BlockHeader) error {
+	// See if the block is known already.
+	id := h.ID()
+	_, exists := cs.dosBlocks[id]
+	if exists {
+		return errDoSBlock
+	}
+
+	// Check if the block is already known.
+	blockMap := tx.Bucket(BlockMap)
+	if blockMap == nil {
+		return errNoBlockMap
+	}
+	if blockMap.Get(id[:]) != nil {
+		return modules.ErrBlockKnown
+	}
+
+	// Check for the parent.
+	parentID := h.ParentID
+	parentBytes := blockMap.Get(parentID[:])
+	if parentBytes == nil {
+		return errOrphan
+	}
+	var parent processedBlock
+	err := cs.marshaler.Unmarshal(parentBytes, &parent)
+	if err != nil {
+		return err
+	}
+
+	// Check that the target of the new block is sufficient.
+	if !checkHeaderTarget(h, parent.ChildTarget) {
+		return modules.ErrBlockUnsolved
+	}
+
+	// TODO: check if the block is a non extending block.
+
+	// Check that the timestamp is not too far in the past to be acceptable.
+	minTimestamp := cs.blockRuleHelper.minimumValidChildTimestamp(blockMap, &parent)
+	if minTimestamp > h.Timestamp {
+		return errEarlyTimestamp
+	}
+
+	// Check if the block is in the extreme future. We make a distinction between
+	// future and extreme future because there is an assumption that by the time
+	// the extreme future arrives, this block will no longer be a part of the
+	// longest fork because it will have been ignored by all of the miners.
+	if h.Timestamp > types.CurrentTimestamp()+types.ExtremeFutureThreshold {
+		return errExtremeFutureTimestamp
+	}
+
+	// Check if the block is in the near future, but too far to be acceptable.
+	if h.Timestamp > types.CurrentTimestamp()+types.FutureThreshold {
+		return errFutureTimestamp
+	}
```

Comment:
no need to have a check for errFutureTimestamp, that's logic that should be handled in acceptBlock only. You may want to add a comment explaining why. The reason is that you want to have the future blocks in memory, so that you can accept them the millisecond that they are no longer too far in the future to be acceptable. Maybe they will have reorged, maybe not. But you want the host to go chase down the block if it finds a header, even if it's a 'future' block.